.NET Standard serialization topic

rpetrusha opened the pull request:
# .NET Standard serialization topic

Shows how to check whether a type in a .NET Standard library is serializable, since the .NET Standard does not define serialization.

## Summary

Includes project files and source code files for VB and C#, and updates the TOC.

Fixes #3249 

## Details

For convenience, I combined the library method in the same file as the console application, which means that it's no longer a library method that targets the .NET Standard. 

The topic is staged at https://review.docs.microsoft.com/en-us/dotnet/standard/serialization/how-to-determine-if-netstandard-object-is-serializable?branch=pr-en-us-3462. Make sure that all checks have passed below, though, to make sure that you're looking at a build that reflects the current content.

## Suggested Reviewers
@mairaw, @BillWagner, @ViktorHofer

Commit:
.NET Standard serialization topic

Commit:
Fixed broken link

ViktorHofer commented:
The code snippet and related links still seem to be broken. I will review again after the code snippet is accessible. Thanks for your work @rpetrusha !!

BillWagner commented on docs/standard/serialization/how-to-determine-if-netstandard-object-is-serializable.md:
```diff
@@ -0,0 +1,37 @@
+---
+# How to: Determine if a .NET Standard object is serializable
+
+The .NET Standard is a specification that defines the types and members that must be present on specific .NET implementations that conform to that version of the standard. However, the .NET Standard does not define whether a type is serializable. The types defined in the .NET Standard Library are not marked with the <xref:System.SerializableAttribute> attribute. Instead, specific .NET implementat ...
```
missing word: "free determine" should be "free to determine"

BillWagner commented on samples/snippets/standard/serialization/is-serializable/csharp/Program.cs:
```diff
@@ -0,0 +1,55 @@
+		    if (UtilityLibrary.IsSerializable(value))
+			{
+				// Serialize the value tuple.
+            var formatter = new BinaryFormatter();
```
indentation on this line is off

BillWagner commented on samples/snippets/standard/serialization/is-serializable/csharp/Program.cs:
```diff
@@ -0,0 +1,55 @@
+				                            FileAccess.Write, FileShare.None);
+				formatter.Serialize(stream, value);
+				stream.Close();
+            // Deserialize the value tuple.
```
These four lines are not indented correctly

BillWagner commented on samples/snippets/standard/serialization/is-serializable/csharp/Program.cs:
```diff
@@ -0,0 +1,55 @@
+    
+    public class UtilityLibrary
+    {
+        public static bool IsSerializable(object obj)
```
Consider making this an extension method:

```csharp
public static bool IsSerializable(this object obj)
```

The example could then be:

```csharp
if (value.IsSerializable())
```

BillWagner commented on samples/snippets/standard/serialization/is-serializable/vb/program.vb:
```diff
@@ -0,0 +1,40 @@
+' <Snippet2>
+Namespace Libraries
+	Public Class UtilityLibrary
+    	Public Shared Function IsSerializable(obj As Object) As Boolean
```
Same comment as the C# sample: consider an extension method

BillWagner reviewed (changes requested):
This looks great Ron.  I'd found a couple small grammar notes and had one other suggestion.

ViktorHofer commented on samples/snippets/standard/serialization/is-serializable/csharp/Program.cs:
```diff
@@ -0,0 +1,55 @@
+				var stream = new FileStream("data.bin", FileMode.Create, 
+				                            FileAccess.Write, FileShare.None);
+				formatter.Serialize(stream, value);
+				stream.Close();
```
Use using around stream instead of manually closing it

ViktorHofer commented on samples/snippets/standard/serialization/is-serializable/csharp/Program.cs:
```diff
@@ -0,0 +1,55 @@
+				stream.Close();
+            // Deserialize the value tuple.
+            var readStream = new FileStream("data.bin", FileMode.Open);
+            var restoredValue = formatter.Deserialize(readStream);
```
object instead of var

ViktorHofer commented on samples/snippets/standard/serialization/is-serializable/csharp/Program.cs:
```diff
@@ -0,0 +1,55 @@
+				formatter.Serialize(stream, value);
+				stream.Close();
+            // Deserialize the value tuple.
+            var readStream = new FileStream("data.bin", FileMode.Open);
```
use using here. same comment as above

ViktorHofer commented on samples/snippets/standard/serialization/is-serializable/csharp/Program.cs:
```diff
@@ -0,0 +1,55 @@
+    {
+        public static bool IsSerializable(object obj)
+        {
+            if (obj == null) 
```
just for reference, the whole expression could be simplified as obj?.GetType()?.IsSerializable ?? false;

ViktorHofer commented on samples/snippets/standard/serialization/is-serializable/csharp/test-serialization.csproj:
```diff
@@ -0,0 +1,11 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <ItemGroup>
```
remove empty ItemGroup

ViktorHofer commented on samples/snippets/standard/serialization/is-serializable/vb/test.serialization.vb.vbproj:
```diff
@@ -0,0 +1,11 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <ItemGroup>
```
same as above. remove empty group.

Commit:
Addressed review comments

rpetrusha commented:
@ViktorHofer, @BillWagner, I've addressed your review comments.

Commit:
Added fxref, corrected bad code reference

rpetrusha commented:
@ViktorHofer and @BillWagner, I've fixed a build error and also added a paragraph to the documentation for System.Type.IsSerializable, if you'd like to take a look.

BillWagner commented:
The updates look good.  Go ahead and :shipit: when the build finishes.

ViktorHofer commented:
Jep. LGTM

mairaw commented on docs/standard/serialization/how-to-determine-if-netstandard-object-is-serializable.md:
```diff
@@ -0,0 +1,37 @@
+---
+title: "How to: Determine if a .NET Standard object is serializable"
+ms.custom: ""
```
nit: no need to add this in new articles

mairaw commented on docs/standard/serialization/how-to-determine-if-netstandard-object-is-serializable.md:
```diff
@@ -0,0 +1,37 @@
+
+You can determine whether an object is serializable at runtime by retrieving the value of the <xref:System.Type.IsSerializable> property of a <xref:System.Type> object that represents that object's type. The following example provides one implementation. It defines an `IsSerializable(Object)` extension method that indicates whether any <xref:System.Object> instance can be serialized.
+
+[!code-csharp[is-a-type-serializable](~/samples/snippets/standard/serialization/is-serializable/csharp/program.cs#2)]
```
this path doesn't look right for snippets unless we want to change it. it should go under the language folders first and then you give a name
so under https://github.com/dotnet/docs/tree/master/samples/snippets/csharp and under https://github.com/dotnet/docs/tree/master/samples/snippets/visualbasic

mairaw commented on docs/standard/serialization/how-to-determine-if-netstandard-object-is-serializable.md:
```diff
@@ -0,0 +1,37 @@
+[!code-csharp[is-a-type-serializable](~/samples/snippets/standard/serialization/is-serializable/csharp/program.cs#2)]
+[!code-vb[is-a-type-serializable](~/samples/snippets/standard/serialization/is-serializable/vb/library.vb#2)]
+
+You can then pass any object to the method to determine whether it can be serialized and deserialized on the current .NET implementation, as the following example shows.
```
nit: end with a colon instead of period

mairaw reviewed (approved):
LGTM but left some comments in case you want to address them

Commit:
Update how-to-determine-if-netstandard-object-is-serializable.md
